scoap3/modules/compliance/compliance.py

```diff
@@ -0,0 +1,226 @@
+    arxiv_array = [
+        a['value'].split(':')[1]
+        for a
+        in obj.data.get('report_numbers', ())
+        if a['source'] == 'arXiv'
+    ]
+    if arxiv_array:
+        return arxiv_array[0]
+    return None
+
+
+def __extract_text_as_extra_data(obj):
+    # fixme extraction shouldn't happen in article_upload?
+    # do extraction only if not done earlier
+    if 'extracted_data' in obj.extra_data:
+        return
+
+    obj.extra_data['extracted_data'] = {}
+    for file in obj.data['_files']:
+        filetype = file['filetype']
+        if filetype in ('pdf', 'pdf/a'):
+            path = ObjectVersion.get(file['bucket'], file['key']).file.uri
+            try:
+                obj.extra_data['extracted_data'][filetype] = extract_text_from_pdf(path).decode('utf-8')
+            except PDFSyntaxError as e:
+                current_app.logger.error('Error while extracting text from pdf with uri %s: %s' % (path, e))
+
+
+
+def __find_regexp(data, patterns):
+    """Finds all matches for given patterns.
+    :return list of all matches.
+    """
+
+    matches = []
+
+    for original_pattern in patterns:
+        # add fuzzing based on pattern length
+        fuzz_i = len(original_pattern) / 11
+
+        # add the surrounding characters too
+        pattern = "(.{0,10}%s.{0,10}){i<=%d}" % (original_pattern, fuzz_i)
+
+        match = regex.findall(pattern, data, regex.IGNORECASE | regex.DOTALL)
+        for m in match:
+            # it can contain a string or tuple of strings if there is multiple match groups
+            if isinstance(m, tuple):
+                matches.append(m[0])
+            else:
+                matches.append(m)
+
+    return matches
+
+
+def __find_regexp_in_pdf(obj, patterns, forbidden_patterns=None):
+    """
+    Finds all matches for given patterns with surrounding characters in all filetypes.
+    Fails only if there are no matches at all or there is a match for a forbidden pattern.
+    :param patterns: iterable of string patterns
+    """
+    __extract_text_as_extra_data(obj)
+
+    ok = True
+    details = []
+
+    for filetype, data in obj.extra_data['extracted_data'].iteritems():
+        # first check for forbidden patterns
+        if forbidden_patterns:
+            forbidden_matches = __find_regexp(data, forbidden_patterns)
+            if forbidden_matches:
+                ok = False
+                details.append('Found forbidden match in %s: "%s"' % (filetype, '", "'.join(set(forbidden_matches))))
+
+        matches = __find_regexp(data, patterns)
+        if not matches:
+            ok = False
+            details.append('Not found in %s' % filetype)
+        else:
+            details.append('Found in %s as: "%s"' % (filetype, '", "'.join(set(matches))))
+
+    return ok, details, None
+
+
+def _files(obj):
+    """check if it has the necessary files: .xml, .pdf, .pdfa """
+
+    file_types = [file['filetype'] for file in obj.data['_files']]
+
+    ok = True
```

Comment:
Maybe it will be better to rename "ok" to something else like "result" - it's strange to assign False to ok